File: detective-game-web/src/game/systems/DialogueSystem.ts
import { SuspectData, QuestionData } from '../data/types';
import { EventBus } from './EventBus';
import { SaveSystem } from './SaveSystem';
import { NotebookSystem } from './NotebookSystem';
import { ClueSystem } from './ClueSystem';
import { CaseLoader } from '../data/caseLoader';

/**
 * DialogueSystem - Manages NPC conversations and question branching
 */
export class DialogueSystem {
  private static currentSuspect: SuspectData | null = null;

  /**
   * Start dialogue with a suspect
   */
  public static startDialogue(suspect: SuspectData): void {
    this.currentSuspect = suspect;

    // Pause game input
    EventBus.emit('input:pause');

    // Open dialogue UI
    EventBus.emit('dialogue:start', {
      suspect,
      questions: this.getAvailableQuestions(suspect),
    });
  }

  /**
   * Get available questions for a suspect
   */
  public static getAvailableQuestions(suspect: SuspectData): QuestionData[] {
    const caseId = CaseLoader.getCurrentCaseId();
    if (!caseId) return suspect.questions;

    const saveData = SaveSystem.load(caseId);
    if (!saveData) return [suspect.questions[0]]; // Only first question initially

    // Get base questions (first question always available)
    const available: QuestionData[] = [];
    const askedIds = saveData.askedQuestions[suspect.id] || [];
    const unlockedIds = saveData.unlockedQuestions[suspect.id] || [];

    suspect.questions.forEach((q, index) => {
      // First question always available
      if (index === 0) {
        available.push(q);
      }
      // Others need to be unlocked
      else if (unlockedIds.includes(q.id)) {
        available.push(q);
      }
    });

    return available;
  }

  /**
   * Check if question has been asked
   */
  public static hasAskedQuestion(
    suspectId: string,
    questionId: string
  ): boolean {
    const caseId = CaseLoader.getCurrentCaseId();
    if (!caseId) return false;

    const saveData = SaveSystem.load(caseId);
    return saveData?.askedQuestions[suspectId]?.includes(questionId) ?? false;
  }

  /**
   * Ask a question
   */
  public static askQuestion(question: QuestionData): void {
    if (!this.currentSuspect) return;

    const caseId = CaseLoader.getCurrentCaseId();
    if (!caseId) return;

    // Save as asked
    SaveSystem.addAskedQuestion(caseId, this.currentSuspect.id, question.id);

    // Add to notebook
    NotebookSystem.addEvent({
      id: `dialogue_${this.currentSuspect.id}_${question.id}_${Date.now()}`,
      timestamp: Date.now(),
      type: 'dialogue',
      title: `Asked ${this.currentSuspect.name}`,
      description: question.text,
    });

    // Process unlocks
    if (question.unlocks) {
      // Unlock new questions
      if (question.unlocks.questions && question.unlocks.questions.length > 0) {
        SaveSystem.unlockQuestions(
          caseId,
          this.currentSuspect.id,
          question.unlocks.questions
        );
        EventBus.emit('toast:show', {
          message: '❓ New question unlocked',
          type: 'info',
        });
      }

      // Unlock clues
      if (question.unlocks.clues && question.unlocks.clues.length > 0) {
        const caseData = CaseLoader.getCurrentCase();
        question.unlocks.clues.forEach((clueId) => {
          const clue = caseData?.clues.find((c) => c.id === clueId);
          if (clue && !ClueSystem.hasClue(clueId)) {
            ClueSystem.collectClue(clue);
          }
        });
      }
    }

    // Emit for UI update
    EventBus.emit('dialogue:question', {
      question,
      answer: question.answer,
    });
  }

  /**
   * End current dialogue
   */
  public static endDialogue(): void {
    this.currentSuspect = null;

    // Resume game input
    EventBus.emit('input:resume');

    // Close dialogue UI
    EventBus.emit('dialogue:end');
  }

  /**
   * Get current suspect
   */
  public static getCurrentSuspect(): SuspectData | null {
    return this.currentSuspect;
  }
}

File: Assets/Scripts/Systems/DialogueSystem.cs
using System.Collections.Generic;
using UnityEngine;
using DetectiveGame.Models;
using DetectiveGame.Core;

namespace DetectiveGame.Systems
{
    /// <summary>
    /// Manages dialogue interactions with suspects, question unlocking, and answer processing.
    /// </summary>
    public class DialogueSystem : MonoBehaviour
    {
        public static DialogueSystem Instance { get; private set; }

        private void Awake()
        {
            if (Instance == null)
            {
                Instance = this;
            }
            else
            {
                Destroy(gameObject);
            }
        }

        /// <summary>
        /// Ask a question to a suspect and process the answer.
        /// </summary>
        public string AskQuestion(string suspectId, string questionId)
        {
            // Check if already asked
            if (SaveService.Instance.IsQuestionAsked(questionId))
            {
                Debug.Log($"[DialogueSystem] Question already asked: {questionId}");
                // Still return the answer for display
            }

            // Find suspect and question
            SuspectData suspect = CaseLoader.Instance.GetSuspect(suspectId);
            if (suspect == null)
            {
                Debug.LogError($"[DialogueSystem] Suspect not found: {suspectId}");
                return null;
            }

            QuestionData question = suspect.questions?.Find(q => q.id == questionId);
            if (question == null)
            {
                Debug.LogError($"[DialogueSystem] Question not found: {questionId}");
                return null;
            }

            // Mark as asked
            SaveService.Instance.AddAskedQuestion(questionId);

            // Add to timeline
            NotebookSystem.Instance?.AddQuestionAskedEntry(suspect, question);

            // Process unlocks
            ProcessUnlocks(suspectId, question);

            // Fire event
            EventBus.TriggerQuestionAsked(suspectId, question);

            Debug.Log($"[DialogueSystem] Asked question to {suspect.name}: {question.text}");
            return question.answer;
        }

        private void ProcessUnlocks(string suspectId, QuestionData question)
        {
            if (question.unlocks == null) return;

            // Unlock new questions
            if (question.unlocks.questions != null)
            {
                foreach (var qId in question.unlocks.questions)
                {
                    if (!SaveService.Instance.CurrentSave.unlockedQuestionIds.Contains(qId))
                    {
                        SaveService.Instance.AddUnlockedQuestion(qId);
                        
                        // Find the question data for the event
                        var suspect = CaseLoader.Instance.GetSuspect(suspectId);
                        var unlockedQ = suspect?.questions?.Find(q => q.id == qId);
                        
                        // Also check other suspects for the unlocked question
                        if (unlockedQ == null)
                        {
                            foreach (var s in CaseLoader.Instance.CurrentCase.suspects)
                            {
                                unlockedQ = s.questions?.Find(q => q.id == qId);
                                if (unlockedQ != null)
                                {
                                    suspectId = s.id;
                                    break;
                                }
                            }
                        }

                        if (unlockedQ != null)
                        {
                            EventBus.TriggerQuestionUnlocked(suspectId, unlockedQ);
                            EventBus.TriggerToast("Câu hỏi mới được mở khóa!", "info");
                        }
                    }
                }
            }

            // Unlock new clues
            if (question.unlocks.clues != null)
            {
                foreach (var clueId in question.unlocks.clues)
                {
                    ClueData clue = CaseLoader.Instance.GetClue(clueId);
                    if (clue != null && !SaveService.Instance.IsClueCollected(clueId))
                    {
                        SaveService.Instance.AddCollectedClue(clueId);
                        EventBus.TriggerClueUnlocked(clue);
                        EventBus.TriggerToast($"Manh mối mới: {clue.name}", "success");
                    }
                }
            }
        }

        /// <summary>
        /// Get available questions for a suspect (initial + unlocked, excluding asked).
        /// </summary>
        public List<QuestionData> GetAvailableQuestions(string suspectId)
        {
            var result = new List<QuestionData>();
            SuspectData suspect = CaseLoader.Instance.GetSuspect(suspectId);
            
            if (suspect?.questions == null) return result;

            foreach (var question in suspect.questions)
            {
                // Check if this is an initial question or has been unlocked
                bool isAvailable = IsQuestionAvailable(question.id);
                
                if (isAvailable)
                {
                    result.Add(question);
                }
            }

            return result;
        }

        /// <summary>
        /// Check if a question is available (initial or unlocked).
        /// </summary>
        public bool IsQuestionAvailable(string questionId)
        {
            // Initial questions (those not requiring unlock) are always available
            // For MVP, we'll mark initial questions by checking if any other question unlocks them
            var save = SaveService.Instance.CurrentSave;
            
            // If explicitly unlocked, it's available
            if (save.unlockedQuestionIds.Contains(questionId))
            {
                return true;
            }

            // Check if this question is unlocked by any other question
            // If not, it's an initial question
            bool isUnlockedByOther = false;
            foreach (var suspect in CaseLoader.Instance.CurrentCase.suspects)
            {
                foreach (var q in suspect.questions)
                {
                    if (q.unlocks?.questions?.Contains(questionId) == true)
                    {
                        isUnlockedByOther = true;
                        break;
                    }
                }
                if (isUnlockedByOther) break;
            }

            // If no other question unlocks this, it's an initial question
            return !isUnlockedByOther;
        }

        /// <summary>
        /// Check if a question has been asked.
        /// </summary>
        public bool IsQuestionAsked(string questionId)
        {
            return SaveService.Instance.IsQuestionAsked(questionId);
        }
    }
}
